embedded-graphics/src/primitives/rounded_rectangle.rs

```diff
@@ -0,0 +1,768 @@
+//! The rounded rectangle primitive.
+
+use crate::{
+    drawable::{Drawable, Pixel},
+    geometry::{Dimensions, Point, Size},
+    pixelcolor::PixelColor,
+    primitives::{
+        ellipse_quadrant::{self, EllipseQuadrant, Quadrant},
+        rectangle::{self, Rectangle},
+        ContainsPoint, Primitive,
+    },
+    style::{PrimitiveStyle, Styled},
+    transform::Transform,
+    DrawTarget,
+};
+
+/// The definition of each corner radius for a rounded rectangle.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct CornerRadii {
+    /// Top left corner radius
+    pub top_left: Size,
+
+    /// Top right corner radius
+    pub top_right: Size,
+
+    /// Bottom right corner radius
+    pub bottom_right: Size,
+
+    /// Bottom left corner radius
+    pub bottom_left: Size,
+}
+
+impl CornerRadii {
+    /// Create a new set of corner radii with all corners having equal values.
+    pub fn new_equal(radius: Size) -> Self {
+        Self {
+            top_left: radius,
+            top_right: radius,
+            bottom_right: radius,
+            bottom_left: radius,
+        }
+    }
+
+    /// Confine corner radii that are too large to a given bounding rectangle
+    fn confine(self, bounding_box: Size) -> Self {
+        // Compute free space for each edge. If there is remaining free space, the call to
+        // checked_sub() will return `None`.
+        let free_space_top =
+            (self.top_left.width + self.top_right.width).checked_sub(bounding_box.width);
+        let free_space_right =
+            (self.top_right.height + self.bottom_right.height).checked_sub(bounding_box.height);
+        let free_space_bottom =
+            (self.bottom_left.width + self.bottom_right.width).checked_sub(bounding_box.width);
+        let free_space_left =
+            (self.top_left.height + self.bottom_left.height).checked_sub(bounding_box.height);
+
+        // Find the largest overlap for all edges that have one
+        let largest_overlap = [
+            free_space_top,
+            free_space_right,
+            free_space_bottom,
+            free_space_left,
+        ]
+        .iter()
+        .filter_map(|s| *s)
+        .max();
+
+        if let Some(largest_overlap) = largest_overlap {
+            // Reduce each corner radius by (largest overlap / 2)
+            let reduce_by = Size::new_equal(largest_overlap / 2);
```

Comment:
When the overlap is e.g. 1px `reduce_by` will be `Size::zero()` and the corners would still overlap. This could be fixed by rounding up: `Size::new_equal((largest_overlap + 1) / 2)`